### PR TITLE
Ensure correct exception thrown for full drive. Include new FileSystemHelpers to check for drive issues before read job

### PR DESCRIPTION
**Pull request by shabtaisharon**

### CODE REVIEW

**Comment by rpmoore** [comment]
There is a build error on this pull request

**Comment by shabtaisharon** [comment]
This is a mono issue, I'll look into that tomorrow morning

**Comment by shabtaisharon** [comment]
updating mono will not help since they dont support the function i'm using on a platform that is not windows
```
		// InternalGet/InternalSet are virtual so that non-Windows platforms which do not share an
		// API between files, mutexes, etc. can override in the subclass and do their own thing.
		internal virtual int InternalGet (SafeHandle handle,
						  AccessControlSections includeSections)
		{
#if MOBILE
			throw new PlatformNotSupportedException ();
#else
			if (Environment.OSVersion.Platform != PlatformID.Win32NT)
				throw new PlatformNotSupportedException ();

			return Win32GetHelper (delegate (SecurityInfos securityInfos,
							 out IntPtr owner, out IntPtr group,
							 out IntPtr dacl, out IntPtr sacl, out IntPtr descriptor)
				{
					return GetSecurityInfo (handle, ResourceType, securityInfos,
								out owner, out group,
								out dacl, out sacl, out descriptor);
				}, includeSections);
#endif
		}
```

File: Ds3/Helpers/FileSystemHelpers.cs [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+        private static ulong DriveFreeBytes(string path)
+        {
+            if (!path.EndsWith("\\"))
+            {
+                path += '\\';
+            }
+
+            var driveInfo  = new DriveInfo(Path.GetPathRoot(path));
+
+            return (ulong)driveInfo.AvailableFreeSpace;
```
Comment (rpmoore) [inline Ds3/Helpers/FileSystemHelpers.cs]: Why do we need the cast to a ulong?

File: Ds3/Helpers/Streams/StreamTranslator.cs [hunk 2]
```diff
@@ -125,7 +126,18 @@ public override void Write(byte[] buffer, int offset, int count)
...
                 {
                     stream.Position = range.Range.Start;
-                    stream.Write(buffer, progress + offset, rangeLength);
+                    try
+                    {
+                        stream.Write(buffer, progress + offset, rangeLength);
+                    }
+                    catch (IOException e)
+                    {
+                        if (e.Message.StartsWith("There is not enough space on the disk"))
```
Comment (rpmoore) [inline Ds3/Helpers/Streams/StreamTranslator.cs]: Is there another way that we can detect this?  The string comparison seems brittle.

File: Ds3/Helpers/FileSystemHelpers.cs (hunk 1 above)
Comment (shabtaisharon) [inline Ds3/Helpers/FileSystemHelpers.cs]: we don't anymore, the old code needed that. i'll change that

File: Ds3/Helpers/Streams/StreamTranslator.cs (hunk 2 above)
Comment (shabtaisharon) [inline Ds3/Helpers/Streams/StreamTranslator.cs]: not that i know of, we only know that this is an IOException, but no way to know that real issue (shrug)

File: Ds3/Helpers/FileSystemHelpers.cs (hunk 1 above)
Comment (shabtaisharon) [inline Ds3/Helpers/FileSystemHelpers.cs]: wait maybe i do have a better way, give me 10 min

Comment (shabtaisharon) [inline Ds3/Helpers/FileSystemHelpers.cs]: ok, it looks like the other way is not better in my opinion
[http://stackoverflow.com/questions/9293227/how-to-check-if-ioexception-is-not-enough-disk-space-exception-type](url)
[https://msdn.microsoft.com/en-us/library/system.exception.hresult(v=vs.110).aspx](url)
@rpmoore let me know what do you think?

File: Ds3/Helpers/Streams/StreamTranslator.cs (hunk 2 above)
Comment (shabtaisharon) [inline Ds3/Helpers/Streams/StreamTranslator.cs]: ok, it looks like the other way is not better in my opinion
http://stackoverflow.com/questions/9293227/how-to-check-if-ioexception-is-not-enough-disk-space-exception-type
https://msdn.microsoft.com/en-us/library/system.exception.hresult(v=vs.110).aspx
@rpmoore let me know what do you think?

**Pull request merged**